gym_handeye/handeye.py

```diff
@@ -0,0 +1,159 @@
+
+        self.reset()
+
+    def step(self, action):
+        """
+        Run one timestep of the environment's dynamics.
+        Accepts an action and returns a tuple (observation, reward, done, info).
+        :param action: an action provided by the environment
+        :return: observation (tuple): agent's observation of the current environment
+            reward (float) : amount of reward returned after previous action
+            done (boolean): whether the episode has ended, in which case further step() calls will return undefined results
+            info (dict): contains auxiliary diagnostic information (helpful for debugging, and sometimes learning)
+        """
+
+        logging.debug('Executing a step, action = {}'.format(action))
+        previous_observation = self._observe()
+        self._take_action(action)
+
+        observation = self._observe()
+        reward = 0
+        episode_over = False
+        if self._should_end_testing(previous_observation, observation):
+            episode_over = True
+
+        return tuple(observation), reward, episode_over, {}
+
+    def reset(self):
+        """
+        Resets the state of the environment and returns an initial observation.
+        Returns:
+        :return: observation (object): the initial observation of the
+            space.
+        """
+        logging.debug('Resetting the environment')
+
+        self.handeye.set_random_positions()
+
+        return tuple(self._observe())
+
+    def render(self, mode='human', close=False):
+        """
+        Renders the environment.
+        :param mode (str): the mode to render with
+        :param close (bool): close all open renderings
+        :return:
+        """
+        if close:
+            return
+
+        logging.debug('Rendering the environment')
+
+        if mode == 'human':
+            outfile = sys.stdout
+            outfile.write("\n")
+
+            j = 0
+            for item in self.handeye.observation:
+                if item not in ['w', 'g', 'b']:
+                    break
+                outfile.write(self._render_element(item))
+                j += 1
+                if j >= self.grid_size:
+                    outfile.write("\n")
+                    j = 0
+        else:
+            super(HandEye, self).render(mode=mode)
+
+    def close(self):
+        return
+
+    @property
+    def unwrapped(self):
+        """
+        Completely unwrap this env.
+        :return:  gym.Env: The base non-wrapped gym.Env instance
+        """
+        return self
+
+    @staticmethod
+    def get_all_possible_actions():
+        return list(range(0, len(ACTION_LOOKUP)))
+
+    def _should_end_testing(self, previous, obs):
+        return (self.test_only_changes and self._no_change_detected(previous, obs)) or (
+                (not self.test_only_changes) and self._change_detected(previous, obs))
+
+    def _no_change_detected(self, previous, current):
```

Comment:
Czy ta metoda nie jest zbędna? Wydaje mi się, że gdzieś w kodzie można po prostu wywołać `not _change_detected`. Drobny refaktor by się przydał